### ISSUE TITLE
[bitnami/minio] bug of non-root containers

### ISSUE THREAD

**Issue by NJ2046** [bitnami/containers#39279]
### Name and Version

bitnami/minio:2021.11.9

### What architecture are you using?

None

### What steps will reproduce the bug?

1. Prepare the following `docker-compose.yml` files
```yaml
version: '2.19'

# Basic function options
x-function: &function
  privileged: true
  restart: always

services:
  minio-service:
    <<: *function
    image: bitnami/minio:2021.11.9
    container_name: minio-service
    volumes:
      - /data/public:/data
    ports:
      - '9000:9000'
      - '9001:9001'
      - '60006:60006'
```
2. `sudo docker-compose up`

### What is the expected behavior?

```console
$ docker-compose up
[+] Building 0.0s (0/0)                                                                                                                                          
[+] Running 2/1
 ✔ Network mtrxs_default    Created                                                                                                                         0.1s 
 ✔ Container minio-service  Created                                                                                                                         0.0s 
Attaching to minio-service
minio-service  |  11:08:54.47 
minio-service  |  11:08:54.47 Welcome to the Bitnami minio container
minio-service  |  11:08:54.47 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-minio
minio-service  |  11:08:54.47 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-minio/issues
minio-service  |  11:08:54.47 
minio-service  |  11:08:54.47 INFO  ==> ** Starting MinIO setup **
minio-service  | minio 11:08:54.49 INFO  ==> Starting MinIO in background...
minio-service  | minio 11:09:04.49 INFO  ==> Adding local Minio host to 'mc' configuration...
minio-service  | minio 11:09:04.52 INFO  ==> Creating default buckets...
minio-service  | minio 11:09:04.56 INFO  ==> Bucket local/matrixsolar already exists, skipping creation.
minio-service  | minio 11:09:04.59 INFO  ==> Stopping MinIO...
minio-service  | 
minio-service  |  11:09:05.66 INFO  ==> ** MinIO setup finished! **
minio-service  | minio 11:09:05.67 INFO  ==> ** Starting MinIO **
minio-service  | 
minio-service  |  You are running an older version of MinIO released 2 years ago 
minio-service  |  Update: Run `mc admin update` 
minio-service  | 
minio-service  | 
minio-service  | API: http://172.19.0.2:9000  http://127.0.0.1:9000     
minio-service  | 
minio-service  | Console: http://172.19.0.2:9001 http://127.0.0.1:9001  
```

### What do you see instead?

```console
$ sudo docker-compose up
[+] Running 12/12
 ✔ minio-service 11 layers [⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿]      0B/0B      Pulled                                                                                                                           5.5s 
   ✔ ad3e435fd20e Pull complete                                                                                                                                                           1.7s 
   ✔ a208cb5e6408 Pull complete                                                                                                                                                           2.0s 
   ✔ 15299a303b5a Pull complete                                                                                                                                                           2.4s 
   ✔ 84fe90b6c5ec Pull complete                                                                                                                                                           2.4s 
   ✔ cd00d051756e Pull complete                                                                                                                                                           2.8s 
   ✔ ecdaff48acba Pull complete                                                                                                                                                           3.9s 
   ✔ a8f891e8cd27 Pull complete                                                                                                                                                           4.0s 
   ✔ 94822a3207a6 Pull complete                                                                                                                                                           4.0s 
   ✔ ed50fcd637c6 Pull complete                                                                                                                                                           4.1s 
   ✔ 40ff4eb48790 Pull complete                                                                                                                                                           4.4s 
   ✔ 5ee449fc199b Pull complete                                                                                                                                                           4.5s 
[+] Building 0.0s (0/0)                                                                                                                                                                        
[+] Running 2/2
 ✔ Network mtrxs_default    Created                                                                                                                                                       0.1s 
 ✔ Container minio-service  Created                                                                                                                                                       0.0s 
Attaching to minio-service
minio-service  |  11:05:21.83 
minio-service  |  11:05:21.83 Welcome to the Bitnami minio container
minio-service  |  11:05:21.83 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-minio
minio-service  |  11:05:21.83 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-minio/issues
minio-service  |  11:05:21.84 
minio-service  |  11:05:21.84 INFO  ==> ** Starting MinIO setup **
minio-service  | /opt/bitnami/scripts/libminio.sh: line 270: /data/.root_user: Permission denied
minio-service exited with code 0
```

### Additional information

In the configuration of Docker, there is a specified mapped directory (volumes). Usually, if the directory does not exist on the host, Minio will help create it recursively. However, with Minio, only the main directory will be created and the directories within the main directory will not be created. Additionally, a permission issue may be reported.

**Comment by NJ2046** [comment]
## What architecture are you using?

* Version used (`minio --version`): bitnami/minio:2021.11.9 or bitnami/minio
* Server setup and configuration: ```40 core cpu 46G mem 128G ssd``` or ```4 core cpu 4G mem 40G```
* Operating System and version (`uname -a`): ```5.4.0-150-generic #167~18.04.1-Ubuntu SMP Wed May 24 00:51:42 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux``` or ```6.1.27-060127-generic #202305051509 SMP PREEMPT_DYNAMIC Fri May  5 15:48:49 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux```

**Comment by carrodher** [comment]
You should grant the proper permissions to the local directory so the service is able to write there. In this case, since it is a non-root container, it should be enough with the local directory belonging to the _root_ group (GID 0)

**Comment by NJ2046** [comment]
Don't you think this is a bug that exists in the dockerfile process?

**Comment by NJ2046** [comment]
```
 In this case, since it is a non-root container
```
please look
```
sudo docker-compose up
```

**Comment by carrodher** [comment]
Could you please elaborate a bit more on why you think this is a bug? In the following blog posts you can find more info about non-root containers and its usage:
- https://docs.bitnami.com/tutorials/work-with-non-root-containers
- https://docs.bitnami.com/kubernetes/faq/configuration/use-non-root/
- https://docs.bitnami.com/tutorials/bitnami-best-practices-hardening-containers/#root-and-non-root-containers

**Comment by NJ2046** [comment]
You're right, thank you

**Comment by NJ2046** [comment]
hi @carrodher ,sorry to bother you again.
```
Using Non-Root Containers as Root Containers
If you wish to run a Bitnami non-root container image as a root container image, you can do so by adding the line user: root right after the image: directive in the container's docker-compose.yml. After making this change, simply restart the container and it will run as the root user with all privileges instead of an unprivileged user.
```
docker compose file
```
  minio-service:
    image: bitnami/minio:2021.11.9
    user: root
    container_name: minio-service
    volumes:
      - /data/public:/data
    ports:
      - '9000:9000'
      - '9001:9001'
      - '60006:60006
```
What do you see instead
```
minio-service  | error: failed switching to "minio": unable to find user minio: no matching entries in passwd file
minio-service exited with code 1
```

**Comment by joancafom** [comment]
Hi @NJ2046 

Yes, it seems there is an issue while running the {{bitnami/minio}} container as a root user instead of the default {{non-root}}, recommended approach. Just to ensure, is this what are you trying to do?

In order to understand your use-case, why do you need to run the container as root?

**Comment by NJ2046** [comment]
```non-root``` is a standardized and safe operation, I agree with that.In this case, they are not concerned about safety but rather about convenience.Because a location for data storage was created in the ```/``` directory, container as root is required,  I tried to persuade them that they didn't care.

**Comment by joancafom** [comment]
Thanks for sharing! Although, as you mentioned, `non-root` is the default strategy followed by the vast majority of bitnami containers, it is true that our documentation mentions that it could be run as `root`. I will create an internal task in order to review this and move the issue to `on hold` until we have updates.

**Comment by gongomgra** [comment]
Hi @NJ2046,

Just a quick note to let you know that we have just released a new image for MinIO both including the fix for this issue and with the latest `2023.08.29` version. You can use the `2023.8.29-debian-11-r0` tag to select the image including the fix

```yaml
# Copyright VMware, Inc.
# SPDX-License-Identifier: APACHE-2.0

version: '2'

services:
  minio:
    user: root
    image: bitnami/minio:2023.8.29-debian-11-r0
    environment:
    - BITNAMI_DEBUG=true
    ports:
      - '9000:9000'
      - '9001:9001'
    volumes:
      - 'minio_data:/bitnami/minio/data'

volumes:
  minio_data:
    driver: local
```

Hope it helps!
